Add Arctic Regional Mesh to COMPASS

mark-petersen opened the pull request:
We had created the Arctic Regional Mesh (ARM) in 2018 with a matlab version. This adds the ability to auto-generate these meshes in COMPASS using our standard python tools, and calling jigsaw from the python package. In particular, a new function `AtlanticPacificGrid` was added to allow different mesh distributions in the two basins.

mark-petersen commented:
The new mesh can be created as follows:
```
cd testing_and_setup/compass/
./list_testcases.py |grep ARM
64: -o ocean -c global_ocean -r ARM60to9 -t init
65: -o ocean -c global_ocean -r ARM60to9 -t spin_up
/setup_testcase.py -f general.config.ocean_turq\
  -m runtime_definitions/mpirun.xml \
  --work_dir \/lustre/scratch4/turquoise/mpeterse/runs/trash/200102_test_ARM \
  -n 64
```
The init was tested for a high resolution region of 9, 6, and 3. Spin-up was copied from another case, but I did not test it and tune the time step and Rayleigh damping.

Full initial condition files cat be found here:
```
cd /lustre/scratch4/turquoise/mpeterse/runs/191230_ARM60to10/ocean/global_ocean/
ARM60to9/init/initial_state/initial_state.nc
ARM60to6/init/initial_state/initial_state.nc
ARM60to3/init/initial_state/initial_state.nc
```

mark-petersen commented:
These plots are created here:
```
cd /lustre/scratch4/turquoise/mpeterse/runs/191230_ARM60to10/ocean/global_ocean/
ls ARM60to9/init/base_mesh/*png
   cellWidthGlobal.png
   cellWidthVsLat.png
```
![cellWidthVsLat](https://user-images.githubusercontent.com/4036816/71824712-73b8cb80-3057-11ea-8dd8-a0e2351a05b7.png)
![cellWidthGlobal](https://user-images.githubusercontent.com/4036816/71824725-79aeac80-3057-11ea-8de9-0fa704b3ec4c.png)

mark-petersen commented on testing_and_setup/compass/ocean/global_ocean/ARM60to9/init/define_base_mesh.py:
```diff
@@ -0,0 +1,40 @@
+    AtlNH = mdt.mergeCellWidthVsLat(lat, 30 * QU1, 9 * QU1, 10.0, 10.0)
+    AtlGrid = mdt.mergeCellWidthVsLat(lat, EC60to30, AtlNH, 0.0, 1.0)
+
+    PacNH = mdt.mergeCellWidthVsLat(lat, 30 * QU1, 9 * QU1, 45.0, 10.0)
```
The high resolution region is controlled by the 9 in the two lines above. The last two arguments are the latitude and width of the transition.

mark-petersen commented on testing_and_setup/compass/ocean/jigsaw_to_MPAS/mesh_definition_tools.py:
```diff
@@ -179,59 +164,42 @@ def RRS_CellWidthVsLat(lat, cellWidthEq, cellWidthPole):
+                    cellWidthOut[j, i] = cellWidthInAtlantic[j]
+            elif lat[j] > 20.0:
+                if (lon[i] > -100.0) & (lon[i] < 35.0):
+                    cellWidthOut[j, i] = cellWidthInAtlantic[j]
```
The definition of the Atlantic region is a simple box with diagonal for Central America. This could be fine-tuned in this section.

mark-petersen commented:
There are vtk files of these meshes for viewing:
culled mesh vtk is here:
```
cd /lustre/scratch4/turquoise/mpeterse/runs/191230_ARM60to10/ocean/global_ocean
ARM60to3/init/culled_mesh/culled_mesh_vtk/  
ARM60to6/init/culled_mesh/culled_mesh_vtk/  
ARM60to9/init/culled_mesh/culled_mesh_vtk/
```
and initial condition is here:
```
cd /lustre/scratch4/turquoise/mpeterse/runs/191230_ARM60to10/ocean/global_ocean
ARM60to9/init/initial_state/initial_state_vtk/
ARM60to6/init/initial_state/initial_state_vtk/
```
where I made the vtk files with:
```
paraview_vtk_field_extractor.py --ignore_time -v allOnCells -f initial_state.nc -o initial_state_vtk -d nVertLevels=0 maxEdges=0
```

milenaveneziani commented:
@mark-petersen: thanks for putting this in!
Just a brief comment to say that we will need to adjust the transition curves, to be able to reproduce what Gennaro did for the ARRM 60to10 and 60to6 grids. I'm talking to him to see where the changes were made in the original matlab script.

mark-petersen commented:
@milenaveneziani if you could tell me the details of what was used before, I will add them in and merge. That is really just the latitude and transition width of the Pacific and Atlantic. I also called it ARM60to9, but I see it is ARM60to10. I can change that too.

milenaveneziani commented:
@mark-petersen,  the grid names are ARRM60to10 and ARRM60to6.

This is matlab script that Gennaro used to generate the 60to10:
```
   ddeg = 1;
   lat = [ -90:ddeg: 90]';
   lon = [-180:ddeg:180]';

   RRS30to10= RRS_CellWidthVsLat(lat,30,10);
   EC60to30 = EC_CellWidthVsLat(lat,30,60,30);
   QU1 = ones(size(lat));

   AtlNH = RRS30to10;
   AtlGrid = mergeCellWidthVsLat(lat, EC60to30, AtlNH, 0, 6);

   PacNH = mergeCellWidthVsLat(lat, 30*QU1,10*QU1, 50, 10);
   PacAC = mergeCellWidthVsLat(lat, PacNH, AtlNH, 50, 10);
   PacGrid = mergeCellWidthVsLat(lat, EC60to30, PacAC, 0, 6);

   cellWidthGlobal = AtlanticPacificGrid(lon, lat, AtlGrid, PacGrid);

   save('-mat','cellWidthVsLatLon.mat','lon','lat','cellWidthGlobal')
```

milenaveneziani commented:
And this is the script he used for the 60to6:
```
   ddeg = 1;
   lat = [ -90:ddeg: 90]';
   lon = [-180:ddeg:180]';

   RRS30to6 = RRS_CellWidthVsLat(lat,30,6);
   EC60to30 = EC_CellWidthVsLat(lat,30,60,30);
   QU1 = ones(size(lat));

   AtlNH = RRS30to6;
   AtlGrid = mergeCellWidthVsLat(lat, EC60to30, AtlNH, 0, 4);

   PacNH = mergeCellWidthVsLat(lat, 30*QU1,6*QU1, 50, 12);
   PacAC = mergeCellWidthVsLat(lat, PacNH, AtlNH, 50, 12);
   PacGrid = mergeCellWidthVsLat(lat, EC60to30, PacAC, 0, 6);

   cellWidthGlobal = AtlanticPacificGrid(lon, lat, AtlGrid, PacGrid);

   save('-mat','cellWidthVsLatLon.mat','lon','lat','cellWidthGlobal')
```

mark-petersen commented:
Perfect, thanks!

milenaveneziani commented:
Here are plots similar to the ones you generated earlier, where the black lines in the first plot are for the 60to10 and the red lines are for the 60to6:

![gridres](https://user-images.githubusercontent.com/8237579/72368887-245c4600-36bc-11ea-919a-94f747fd1eb0.png)
![cellSizeGlobal_60to10](https://user-images.githubusercontent.com/8237579/72368971-55d51180-36bc-11ea-9b06-2c47c5922ed3.png)
![cellSizeGlobal_60to6](https://user-images.githubusercontent.com/8237579/72368884-23c3af80-36bc-11ea-997d-d88e2f70e56b.png)

mark-petersen commented:
rebased, removed ocean11 references, added @milena's alterations above. The init process runs through, and here are plots for comparison:
![image](https://user-images.githubusercontent.com/4036816/73609212-baa0cf00-45cb-11ea-8e22-b4ff4d60574e.png)

mark-petersen commented:
I tested the generation of both meshes, and they are here on IC:
```
/lustre/scratch4/turquoise/mpeterse/runs/200202_ARM60to6/ocean/global_ocean/ARM60to6/init
/lustre/scratch4/turquoise/mpeterse/runs/200202_ARM60to10/ocean/global_ocean/ARM60to10/init
```
I think this is sufficiently tested to merge in for future use.

milenaveneziani commented:
Thanks @mark-petersen for doing this!
